### PR TITLE
AffineGrid, RandAffineGrid, Resample, Affine, RandAffine, Affined, RandAffined, Rand2dElastic, Rand3dElastic, Rand2dElasticd, Rand3dElasticd

### PR DESCRIPTION
**Pull request by rijobro**
Partially addresses: https://github.com/Project-MONAI/MONAI/issues/2917.

### Description
Torch compatibility for `AffineGrid`, `RandAffineGrid` and `Resample`, `Affine`, `RandAffine`, `Affined`, `RandAffined`, `Rand2dElastic`, `Rand3dElastic`, `Rand2dElasticd`, `Rand3dElasticd`.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by rijobro** [comment]
This PR removes `as_tensor_output` (I plan on removing these from all transforms). Should I re-add them and mark them as deprecated?

**Comment by rijobro** [comment]
@wyli [this flake8 error](https://github.com/Project-MONAI/MONAI/pull/2922/checks?check_run_id=3587468777#step:7:115) seems unrelated to my changes. Any ideas?

**Comment by wyli** [comment]
> @wyli [this flake8 error](https://github.com/Project-MONAI/MONAI/pull/2922/checks?check_run_id=3587468777#step:7:115) seems unrelated to my changes. Any ideas?

just saw that as well, https://github.com/Project-MONAI/MONAI/pull/2940 fixes it, could you please review that first?

File: monai/transforms/spatial/array.py [hunk 1]
```diff
@@ -1245,11 +1238,15 @@ def __call__(self, spatial_size: Sequence[int]):
 
 
 class Resample(Transform):
+
+    backend = [TransformBackends.TORCH]
```
Comment (Nic-Ma) [inline monai/transforms/spatial/array.py:1242]: Maybe you can add the support for numpy input for this transform directly?

Thanks.

**Review by Nic-Ma — APPROVED** [review]

**Comment by wyli** [comment]
any idea what's the root cause of this one https://github.com/Project-MONAI/MONAI/runs/3599597976?
I run `python -m tests.test_invertd`, works fine on both mac os and ubuntu

File: monai/transforms/spatial/array.py [hunk 2]
```diff
@@ -1084,17 +1083,13 @@ def __call__(
         else:
             affine = self.affine
 
-        if isinstance(affine, np.ndarray):
-            affine = torch.as_tensor(np.ascontiguousarray(affine))
+        if self.device not in (None, torch.device("cpu"), "cpu"):
```
Comment (Nic-Ma) [inline monai/transforms/spatial/array.py:1086]: Hi @rijobro ,

May I know why you check the device type here? I don't quite understand the logic.

Thanks in advance.

**Pull request merged**